### PR TITLE
[Merged by Bors] - add a `SceneBundle` to spawn a scene

### PR DESCRIPTION
**Pull request by mockersf**
# Objective

- Spawning a scene is handled as a special case with a command `spawn_scene` that takes an handle but doesn't let you specify anything else. This is the only handle that works that way.
- Workaround for this have been to add the `spawn_scene` on `ChildBuilder` to be able to specify transform of parent, or to make the `SceneSpawner` available to be able to select entities from a scene by their instance id

## Solution

Add a bundle
```rust
pub struct SceneBundle {
    pub scene: Handle<Scene>,
    pub transform: Transform,
    pub global_transform: GlobalTransform,
    pub instance_id: Option<InstanceId>,
}
```

and instead of 
```rust
commands.spawn_scene(asset_server.load("models/FlightHelmet/FlightHelmet.gltf#Scene0"));
```
you can do
```rust
commands.spawn_bundle(SceneBundle {
    scene: asset_server.load("models/FlightHelmet/FlightHelmet.gltf#Scene0"),
    ..Default::default()
});
```

The scene will be spawned as a child of the entity with the `SceneBundle`

~I would like to remove the command `spawn_scene` in favor of this bundle but didn't do it yet to get feedback first~

### CODE REVIEW

**Review by alice-i-cecile — COMMENTED** [review]
I am pretty firmly in favor of this change, and removing spawn_scene as well. It reduces their mystery significantly, and increases configurability while shrinking our API surface.

Could you modify https://github.com/bevyengine/bevy/blob/main/examples/scene/scene.rs to reflect the new API?

**Comment by NathanSWard** [comment]
Seeing this really makes me want to have `TransformBundle` haha :)

File: crates/bevy_scene/src/bundle.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+#[allow(clippy::type_complexity)]
+pub fn scene_bundle_spawner(
+    mut scene_to_spawn: Query<
+        (Entity, &Handle<Scene>, &mut Option<InstanceId>),
+        Changed<Handle<Scene>>,
+    >,
+    mut scene_spawner: ResMut<SceneSpawner>,
+) {
+    for (entity, scene, mut instance) in scene_to_spawn.iter_mut() {
```
Comment (NathanSWard) [inline crates/bevy_scene/src/bundle.rs:25]: I have a preference to use `for_each_mut` since it's slightly more performant.

File: crates/bevy_scene/src/bundle.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        if let Some(id) = *instance {
```
Comment (NathanSWard) [inline crates/bevy_scene/src/bundle.rs:26]: To avoid triggering change detection twice here, you can use `Option::replace`
```rust
if let Some(id) = instance.replace(..) {
    //..
}

**Review by NathanSWard — CHANGES_REQUESTED** [review]
Look pretty good :)

File: crates/bevy_scene/src/bundle.rs (hunk 1 above)
Comment (DJMcNab) [reply crates/bevy_scene/src/bundle.rs:25]: We believe that opt in `Component` will fix the performance difference, (hopefully by 0.6?)

**Comment by Davier** [comment]
This new API doesn't seem to allow spawning a scene without a parent entity, which was the default before. Is that alright / on purpose? I don't see an obvious way to add that option on top of it.
Also, I believe it allows to spawn the scene as a child of an existing entity, but inserting the `SceneBundle` into that entity will overwrite its `Transform`. That might deserve an example and/or a warning in the docs.

I like the idea of storing some scene data as components though. I'd be in favor of gradually removing some of the fields of `SceneSpawner` (maybe only leave what handles `DynamicScene`s).

File: crates/bevy_scene/src/bundle.rs (hunk 1 above)
Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs:25]: I'd prefer to keep the `for .. in ..` loop as it's what used through Bevy

**Comment by mockersf** [comment]
> This new API doesn't seem to allow spawning a scene without a parent entity, which was the default before. Is that alright / on purpose? I don't see an obvious way to add that option on top of it.

There are two ways to insert a scene for the moment:
* through the `SceneSpawner` resource. it queues the scene to be spawned at the next run of the `scene_spawner_system` in stage `PreUpdate`
* through the `SpawnScene` command. it actually just calls the `SceneSpawner` resource when the command is executed

I want to remove the `SpawnScene` command, and replace it by the `SceneBundle`. Advanced use would still be possible through the `SceneSpawner` resource.

> Also, I believe it allows to spawn the scene as a child of an existing entity, but inserting the `SceneBundle` into that entity will overwrite its `Transform`. That might deserve an example and/or a warning in the docs.

That's how all the bundles with a transform component work already... If you insert a `SpriteBundle` in an entity with a `Transform` component, it will overwrite the existing one

**Comment by mockersf** [comment]
* added `DynamcSceneBundle` for dynamic scene
* removed `SpawnScene` command
* depends on #2410 to work properly

File: crates/bevy_transform/src/components/parent.rs [hunk 3]
```diff
@@ -22,7 +22,11 @@ impl FromWorld for Parent {
 
 impl MapEntities for Parent {
     fn map_entities(&mut self, entity_map: &EntityMap) -> Result<(), MapEntitiesError> {
-        self.0 = entity_map.get(self.0)?;
+        // Parent of an entity in the new world can be in outside world, in which case it
+        // should not be mapped.
+        if let Ok(mapped_entity) = entity_map.get(self.0) {
+            self.0 = mapped_entity;
+        }
```
Comment (Davier) [inline crates/bevy_transform/src/components/parent.rs:29]: Why is this special cased compared to other components that map entities?
Maybe we could always avoid unwrapping at the call sites of `map_entities` and log errors instead?

**Comment by cart** [comment]
Just chiming in to say that I do think something like this is a good direction for the scene system. It anchors scenes more directly to the ECS and makes the common case of transforming scenes / adding components to the "main scene entity" more natural + consistent. I'll take a deeper look at this after the 0.6 release.

File: crates/bevy_scene/src/bundle.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+use crate::{DynamicScene, InstanceId, Scene, SceneSpawner};
+
+#[derive(Default, Component)]
+pub struct HasSceneInstance(Option<InstanceId>);
+
+#[derive(Default, Bundle)]
+pub struct SceneBundle {
+    pub scene: Handle<Scene>,
+    pub instance_id: HasSceneInstance,
+    pub transform: Transform,
```
Comment (IceSentry) [inline crates/bevy_scene/src/bundle.rs:20]: Should this use the TransformBundle instead?

Comment (alice-i-cecile) [reply crates/bevy_scene/src/bundle.rs:20]: No: we should try to avoid nested bundles where feasible.

File: crates/bevy_scene/src/bundle.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+};
+use bevy_transform::components::{GlobalTransform, Transform};
+
+use crate::{DynamicScene, InstanceId, Scene, SceneSpawner};
+
+#[derive(Default, Component)]
+pub struct HasSceneInstance(Option<InstanceId>);
+
+#[derive(Default, Bundle)]
+pub struct SceneBundle {
```
Comment (alice-i-cecile) [inline crates/bevy_scene/src/bundle.rs:17]: Basic docs please.

File: crates/bevy_scene/src/bundle.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            .0
+            .replace(scene_spawner.spawn_as_child(scene.clone(), entity))
+        {
+            scene_spawner.despawn_instance(id);
+        }
+    }
+}
+
+#[derive(Default, Bundle)]
+pub struct DynamicSceneBundle {
```
Comment (alice-i-cecile) [inline crates/bevy_scene/src/bundle.rs:43]: Basic docs

File: crates/bevy_scene/src/bundle.rs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    entity::Entity,
+    prelude::{Changed, Component},
+    system::Query,
+};
+use bevy_transform::components::{GlobalTransform, Transform};
+
+use crate::{DynamicScene, InstanceId, Scene, SceneSpawner};
+
+#[derive(Default, Component)]
+pub struct HasSceneInstance(Option<InstanceId>);
```
Comment (alice-i-cecile) [inline crates/bevy_scene/src/bundle.rs:14]: In the code below, this field is referenced repeatedly as .0 and it's hard to follow. Can we change this to a named struct, with an `id` field?

Comment (alice-i-cecile) [inline crates/bevy_scene/src/bundle.rs:14]: This type is pub, but the field is unreadable. We should probably pub the field IMO to make it more useful, or at least write a getter.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Docs and nits. The core idea is good and well-executed.

File: crates/bevy_scene/src/bundle.rs (hunk 7 above)
Comment (jakobhellermann) [inline crates/bevy_scene/src/bundle.rs:14]: We can do this either by having a wrapper component around an `HasSceneInstance(Option<InstanceId>)` or have a `SceneInstance(InstanceId)` component that is not in the bundle, but gets added when the scene is spawned.

The latter would have the advantage that it is easier to query for the instance id (just `Query<&SceneInstance>`).

File: crates/bevy_scene/src/scene_spawner.rs [hunk 8]
```diff
@@ -73,26 +88,37 @@ impl SceneSpawner {
...
     pub fn despawn_sync(
         &mut self,
         world: &mut World,
         scene_handle: Handle<DynamicScene>,
     ) -> Result<(), SceneSpawnError> {
-        if let Some(instance_ids) = self.spawned_dynamic_scenes.get(&scene_handle) {
+        if let Some(instance_ids) = self.spawned_dynamic_scenes.remove(&scene_handle) {
             for instance_id in instance_ids {
-                if let Some(instance) = self.spawned_instances.get(instance_id) {
+                if let Some(instance) = self.spawned_instances.get(&instance_id) {
```
Comment (jakobhellermann) [inline crates/bevy_scene/src/scene_spawner.rs:102]: isn't this redundant if `despawn_instance_sync` is called below?

**Review by jakobhellermann — APPROVED** [review]

File: crates/bevy_scene/src/bundle.rs (hunk 7 above)
Comment (alice-i-cecile) [reply crates/bevy_scene/src/bundle.rs:14]: I like this strategy more: it's also a lot simpler to understand and could be private.

Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs:18]: implemented `Deref` and `DerefMut` for it

Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs:18]: implemented `Deref` and `DerefMut` for it

Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs:18]: having the component on the bundle allows for easier discoverability that it exists, but at the cost of it being an option. I think the improved discoverability is worth it

File: crates/bevy_scene/src/bundle.rs (hunk 5 above)
Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs:36]: done

File: crates/bevy_scene/src/bundle.rs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+#[derive(Default, Bundle)]
+pub struct SceneBundle {
+    pub scene: Handle<Scene>,
+    pub instance_id: HasSceneInstance,
+    pub transform: Transform,
+    pub global_transform: GlobalTransform,
+}
+
+#[allow(clippy::type_complexity)]
+pub fn scene_bundle_spawner(
```
[the remark this answers is not in the document]
Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs:47]: done

File: crates/bevy_scene/src/bundle.rs (hunk 6 above)
Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs:63]: done

File: crates/bevy_scene/src/scene_spawner.rs (hunk 8 above)
Comment (mockersf) [reply crates/bevy_scene/src/scene_spawner.rs]: indeed!

File: crates/bevy_scene/src/bundle.rs (hunk 7 above)
Comment (cart) [reply crates/bevy_scene/src/bundle.rs:18]: I think in this case being able to easily query for "ready scenes" is the better UX. Otherwise we're putting the burden on the user to filter through all "unready scenes". Discoverability is an issue, but thats also true for a number of marker components, and I think we've largely decided that its a ["documentation and editor" issue](https://github.com/bevyengine/bevy/issues/3833)?

**Comment by cart** [comment]
I think my biggest issue with this is that it "formalizes" (and actively encourages) the "Scenes live under a parent node" pattern, which only really exists because we have no good way to "override" scene components.

I think in an ideal world I think I would make the following changes:
1. Scenes must have a single "root" entity
2. Spawning a scene supports ergonomically "overriding" the default components (especially at the root)

I think these are pretty "fundamental" changes and they should probably be tackled holistically with the "next" scene iteration (whatever that ends up being).

I do like the `SceneInstance` component: this feels like the right general direction for deferred scene spawning (although I think I agree that it should be `SceneInstance(InstanceId)` instead of using an option). 

The extra "top level" node makes it harder to design scenes as logical units. You no longer have "the player scene". You have "the spawned player scene plus the additional top level player node". If you add a Player component to your "player scene", you can no longer query for `&Transform, With<Player>` to get the position. You need to know to query for the Player, then step up a level in the hierarchy to get the Transform position.

Yes, the current system encourages this pattern, but idk if doubling down on it while we wait for the ideal future impl is the move.

**Comment by cart** [comment]
That being said, if we do (collectively) decide to go down this path in the short term, I think this PR is a solid impl / ready to go.

File: crates/bevy_scene/src/bundle.rs (hunk 7 above)
Comment (alice-i-cecile) [reply crates/bevy_scene/src/bundle.rs:18]: I agree with Cart's position here.

**Comment by mockersf** [comment]
> I think my biggest issue with this is that it "formalizes" (and actively encourages) the "Scenes live under a parent node" pattern, which only really exists because we have no good way to "override" scene components.

There are currently two ways to spawn a scene:
* using the `SceneSpawner` resource, which spawn the scene directly in the world and gives the instance id
* using the `spawn_scene` command, which provide very little control on the scene spawned unless it's part of a hierarchy

This PR removes the command, but the `SceneSpawner` is still available. The experience of spawning a scene at a given point in the world (with an exiting transform) is greatly improved by making it a bundle. 

I moved more examples to use the `SceneSpawner`, when not needing to apply a transform from the start to a scene.

File: crates/bevy_scene/src/bundle.rs (hunk 7 above)
Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs]: move to a not option 👍

File: crates/bevy_scene/src/bundle.rs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    fn deref_mut(&mut self) -> &mut <Self as Deref>::Target {
+        &mut self.0
+    }
+}
+
+/// A component bundle for a [`Scene`] root.
+///
+/// Once the scene is spawned, the entity will have a [`SceneInstance`] commponent.
+#[derive(Default, Bundle)]
+pub struct SceneBundle {
```
Comment (nicopap) [inline crates/bevy_scene/src/bundle.rs:38]: Say I want to add the `Scene` to some entity while keeping the existing transform. I could just do a `commands.entity(to_update).insert(scene_handle)` right? It might be worthwhile adding that info to the `Scene` doc. Maybe here too, but I don't think it is as necessary.

**Review by nicopap — APPROVED** [review]

File: crates/bevy_scene/src/bundle.rs (hunk 10 above)
Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs:38]: Yes, but that's the same for all bundles: if you want to add one to an existing entity without overriding the existing value of component, you can't add the bundle but need to add only the new components

Comment (alice-i-cecile) [reply crates/bevy_scene/src/bundle.rs:38]: Agreed; this should be a seperate bit of functionality that's universal for all bundles.

Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs:38]: added some basic doc 👍

**Review by alice-i-cecile — APPROVED** [review]

**Comment by cart** [comment]
Just resolved merge conflicts and removed the remaining instances of `SceneSpawner` in favor of SceneBundle (with the exception of `scene_viewer`, which would need some more thought to properly adapt it).

A slightly controversial change: I adapted `update_gltf_scene` to use SceneBundle and hierarchy navigation instead of querying instance entities in SceneSpawner. I think the "hierarchy navigation" method of scene entity access is more generically useful: usually users will care about specific entities in a scene, and hierarchy navigation is the better way to find those entities (imo). This also had the added benefit of making the example less stateful / complicated (and it will be even simpler once we add a built in ergonomic way to iterate an entire hierarchy).

I _am_ warming up to this as an interim short-term fix, despite it encouraging structures and patterns that will probably change in the near future. The migration of examples was simple enough (and the reduction in complexity for non-trivial examples big enough) to justify this as an intermediate step.

File: crates/bevy_scene/src/bundle.rs [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+use bevy_transform::components::{GlobalTransform, Transform};
+
+use crate::{DynamicScene, InstanceId, Scene, SceneSpawner};
+
+/// [`InstanceId`] of a spawned scene. It can be used with the [`SceneSpawner`] to
+/// interact with the spawned scene.
+#[derive(Component)]
+pub struct SceneInstance(InstanceId);
+
+impl Deref for SceneInstance {
```
Comment (alice-i-cecile) [inline crates/bevy_scene/src/bundle.rs:20]: This should be replaced with the new Deref and DerefMut derives.

File: crates/bevy_scene/src/bundle.rs [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+impl DerefMut for SceneInstance {
+    fn deref_mut(&mut self) -> &mut <Self as Deref>::Target {
+        &mut self.0
+    }
+}
+
+/// A component bundle for a [`Scene`] root.
+///
+/// Once the scene is spawned, the entity will have a [`SceneInstance`] component.
```
Comment (alice-i-cecile) [inline crates/bevy_scene/src/bundle.rs:36]: The PR description says the scene will be added as a Child. Is that still accurate? If so, it should be documented here and on `DynamicSceneBundle`.

File: crates/bevy_scene/src/bundle.rs [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+#[derive(Default, Bundle)]
+pub struct SceneBundle {
+    /// Handle to the scene to spawn
+    pub scene: Handle<Scene>,
+    pub transform: Transform,
+    pub global_transform: GlobalTransform,
+}
+
+/// System that will spawn scenes from [`SceneBundle`].
+pub fn scene_spawner(
```
Comment (alice-i-cecile) [inline crates/bevy_scene/src/bundle.rs:46]: Can we merge this with `dynamic_scene_spawner`? I don't think these make a ton of sense as distinct systems.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Some small cleanup comments.

File: examples/3d/update_gltf_scene.rs [hunk 14]
```diff
@@ -37,56 +25,49 @@ fn setup(
...
-    let instance_id =
-        scene_spawner.spawn(asset_server.load("models/FlightHelmet/FlightHelmet.gltf#Scene0"));
-    scene_instance.0 = Some(instance_id);
+    commands
+        .spawn_bundle(SceneBundle {
+            scene: asset_server.load("models/FlightHelmet/FlightHelmet.gltf#Scene0"),
+            ..default()
+        })
+        .insert(MovedScene);
 }
```
Comment (Shatur) [inline examples/3d/update_gltf_scene.rs:42]: Looks like this comment is no longer relevant?

File: examples/3d/load_gltf.rs [hunk 15]
```diff
@@ -14,8 +14,12 @@ fn main() {
...
 }
 
-fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
-    commands.spawn_scene(asset_server.load("models/FlightHelmet/FlightHelmet.gltf#Scene0"));
+fn setup(
+    mut commands: Commands,
+    asset_server: Res<AssetServer>,
+    mut scene_spawner: ResMut<SceneSpawner>,
+) {
+    scene_spawner.spawn(asset_server.load("models/FlightHelmet/FlightHelmet.gltf#Scene0"));
```
Comment (Shatur) [inline examples/3d/load_gltf.rs:22]: Can we use a bundle here instead of `SceneSpawner`?

File: examples/asset/hot_asset_reloading.rs [hunk 16]
```diff
@@ -16,15 +16,19 @@ fn main() {
...
+) {
     // Load our mesh:
     let scene_handle = asset_server.load("models/monkey/Monkey.gltf#Scene0");
 
     // Any changes to the mesh will be reloaded automatically! Try making a change to Monkey.gltf.
     // You should see the changes immediately show up in your app.
 
     // mesh
-    commands.spawn_scene(scene_handle);
+    scene_spawner.spawn(scene_handle);
```
Comment (Shatur) [inline examples/asset/hot_asset_reloading.rs:31]: And here:

File: crates/bevy_scene/src/bundle.rs (hunk 12 above)
Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs:36]: That's what I meant by `root`. I tried to improve the doc 👍

File: crates/bevy_scene/src/bundle.rs (hunk 13 above)
Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs:46]: they can run in parallel, it depends if we want to promote many small systems or less bigger systems?

File: examples/3d/update_gltf_scene.rs (hunk 14 above)
Comment (mockersf) [reply examples/3d/update_gltf_scene.rs]: updated 👍

File: examples/3d/load_gltf.rs (hunk 15 above)
Comment (mockersf) [reply examples/3d/load_gltf.rs]: switched 👍

File: examples/asset/hot_asset_reloading.rs (hunk 16 above)
Comment (mockersf) [reply examples/asset/hot_asset_reloading.rs]: switched too 👍

File: crates/bevy_scene/src/bundle.rs (hunk 13 above)
Comment (alice-i-cecile) [reply crates/bevy_scene/src/bundle.rs:47]: Hmm. I think that in this case it's clearer to merge them, and the overhead will outweigh the parallelism because of how rarely active these will be.

Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs:59]: merged!

File: crates/bevy_scene/src/bundle.rs (hunk 11 above)
Comment (cart) [reply crates/bevy_scene/src/bundle.rs:20]: I think I disagree:
* Deref/DerefMut derives are pretty "small" and uncomplex. I highly doubt they meaningfully affect compile times
* I think this falls under the "engine code should look like user code" category. If we aren't willing to use our abstractions in internals, we shouldn't be recommending them to users (not a hard and fast rule, but a good rule of thumb imo).
* I don't think adding a dependency on bevy_derive meaningfully affects our dependency tree here.

Comment (cart) [reply crates/bevy_scene/src/bundle.rs:20]: Also the less code we manually write out, the better.

Comment (mockersf) [reply crates/bevy_scene/src/bundle.rs]: done
